### PR TITLE
Add documentation for Serving probes

### PR DESCRIPTION
**Pull request by ReToCode**
## Proposed Changes
- Adds documentation for https://github.com/knative/serving/pull/14853

/assign @dprotaso 
/assign @skonto

### CODE REVIEW

**Comment by ReToCode** [comment]
@dprotaso gentle ping.

File: docs/serving/services/configure-probing.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+One reason for this is that Knative tries to minimize cold-start time and thus is probing
+in a vastly higher interval than Kubernetes does. 
+
+The general probing architecture looks like this:
+
+![probing-overview](./probes-overview.drawio.svg)
+
+* Users can optionally define Readiness and/or Liveness probes in the `KnativeService` CR.
+* The Liveness probes are directly executed by the Kubelet against the according container.
+* Readiness probes, on the other hand, are rewritten by Knative to be executed by the Queue-Proxy container to allow more aggressive probing and faster cold-start times.
```
Comment (skonto) [inline docs/serving/services/configure-probing.md:15]: Could you explain a bit more the aggressive probing and who does probing eg. activator, ingress (at least as abackground info this info valuable for regular users in order to understand a bit of the internals and during debugging).

File: docs/serving/services/configure-probing.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+Supported probe types are:
+
+* httpGet
+* tcpSocket
+* exec
+* grpc
+
+
+!!! note
+    Be aware that Knative also does some defaulting and additional validation to make aggressive probing work.
```
Comment (skonto) [inline docs/serving/services/configure-probing.md:77]: Maybe add the defaulting info here. This section I think could become a good reference for all probing info wdyth?

File: docs/serving/services/configure-probing.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+## General understanding of Knative Probing
+
+It is important to note that Knative probing is different from Kubernetes probing. 
+One reason for this is that Knative tries to minimize cold-start time and thus is probing
+in a vastly higher interval than Kubernetes does. 
+
+The general probing architecture looks like this:
+
+![probing-overview](./probes-overview.drawio.svg)
```
Comment (skonto) [inline docs/serving/services/configure-probing.md:11]: Might be useful to describe a bit of the details like QP will hit each of the endpoints one by one and how the probe headers are passed. How the re-write looks like?

**Comment by skonto** [comment]
/lgtm. 
/hold
It is an opportunity to expand the whole section a bit with more info to educate users imho.

File: docs/serving/services/configure-probing.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+* Users can optionally define Readiness and/or Liveness probes in the `KnativeService` CR.
+* The Liveness probes are directly executed by the Kubelet against the according container.
+* Readiness probes, on the other hand, are rewritten by Knative to be executed by the Queue-Proxy container to allow more aggressive probing and faster cold-start times.
+* Knative will define a default Readiness probe for the primary user container when no probe is defined by the user. It will check for a TCP socket on the traffic port of the Knative Service.
+* Knative will also define a Readiness probe for the Queue-Proxy container itself. Queue-Proxy's health endpoint aggregates all results from it's the rewritten Readiness probes for all user containers (primary + sidecars).
```
Comment (skonto) [inline docs/serving/services/configure-probing.md:17]: Might be useful to describe a bit of the details like QP will hit each of the endpoints one by one and how the probe headers are passed. How the re-write looks like?

**Comment by ReToCode** [comment]
> It is an opportunity to expand the whole section a bit with more info to educate users imho.

Good point, I'll see what I can add without making it too complex here (e.g. not all internas are really relevant to the "casual" user). But we maybe can also link the FT document that has a whole table of combinations as well?

**Comment by skonto** [comment]
> But we maybe can also link the FT document that has a whole table of combinations as well?

Do we do this elsewhere? I am afraid we might change the implementation without changing the ft doc (happened before) because ft doc has its own lifecycle. So I would say let's keep one place where info is updated. wdyth?

**Comment by ReToCode** [comment]
I added more details in the docs. But I don't think this is the place to really put all the details about headers, how the components handle those and the exact rewriting. This documentation is focussing on the end-user and not Knative contributors. I think for those folks, the FTs + the actual code should be enough (at least it was for me working on this). An additional documentation is probably just going to be outdated at some point anyway.

File: docs/serving/services/configure-probing.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+* The Liveness probes are directly executed by the Kubelet against the according container.
+* Readiness probes, on the other hand, are rewritten by Knative to be executed by the Queue-Proxy container.
+* Knative does probing from in places (e.g. Activator, net-* controller and from Queue-Proxy), to make sure the whole network stack is configured and ready. Compare to vanilla Kubernetes, Knative uses faster (called aggressive probing) probing interval to shorten the cold-start times when a Pod is already up and running while Kubernetes itself has not yet reflected that readiness.
+* Knative will define a default Readiness probe for the primary user container when no probe is defined by the user. It will check for a TCP socket on the traffic port of the Knative Service.
+* Knative will also define a Readiness probe for the Queue-Proxy container itself. Queue-Proxy's health endpoint aggregates all results from it's the rewritten Readiness probes for all user containers (primary + sidecars). For the aggregated status, Queue-Proxy will call each container's Readiness probe in parallel, wait for their response (or timeout) and report an aggregated result back to Kubernetes.
```
Comment (skonto) [inline docs/serving/services/configure-probing.md:18]: "from it's rewritten Readiness probes"

File: docs/serving/services/configure-probing.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+in a vastly higher interval than Kubernetes does. 
+
+The general probing architecture looks like this:
+
+![probing-overview](./probes-overview.drawio.svg)
+
+* Users can optionally define Readiness and/or Liveness probes in the `KnativeService` CR.
+* The Liveness probes are directly executed by the Kubelet against the according container.
+* Readiness probes, on the other hand, are rewritten by Knative to be executed by the Queue-Proxy container.
+* Knative does probing from in places (e.g. Activator, net-* controller and from Queue-Proxy), to make sure the whole network stack is configured and ready. Compare to vanilla Kubernetes, Knative uses faster (called aggressive probing) probing interval to shorten the cold-start times when a Pod is already up and running while Kubernetes itself has not yet reflected that readiness.
```
Comment (skonto) [inline docs/serving/services/configure-probing.md:16]: /s/Compare/Compared

File: docs/serving/services/configure-probing.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+* tcpSocket
+* exec
+* grpc
+
+
+!!! note
+    Be aware that Knative also does some defaulting (checking readiness on the traffic port using a HTTP check) and additional validation to make aggressive probing work.
+
+!!! warning
+    As the Queue-Proxy container does not rewrite or check defined Liveness probes, it is important to know that Kubernetes can and will restart specific containers once a Liveness probe fails. Make sure to also include the same check that you define as a Liveness probe as a Readiness probe to make sure Knative is aware of the failing container in the Pod. Otherwise, it is possible that you see connection errors during the restart of a container caused by the Liveness probe failure.
```
Comment (skonto) [inline docs/serving/services/configure-probing.md:81]: Suggestion: "Otherwise, you may see.."

File: docs/serving/services/configure-probing.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+Supported probe types are:
+
+* httpGet
+* tcpSocket
+* exec
+* grpc
+
+
+!!! note
+    Be aware that Knative also does some defaulting (checking readiness on the traffic port using a HTTP check) and additional validation to make aggressive probing work.
```
Comment (skonto) [inline docs/serving/services/configure-probing.md:78]: "an HTTP check"

File: docs/serving/services/configure-probing.md (hunk 6 above)
Comment (skonto) [inline docs/serving/services/configure-probing.md:16]: Add a comma, "(e.g. Activator, net-* controller, and from Queue-Proxy)

**Comment by skonto** [comment]
Thanks @ReToCode for the updates some I added some minor comments. LGTM, I will wait @dprotaso for his review before merging.

**Comment by ReToCode** [comment]
@skonto updated. I think we can go ahead without Dave here, as we can always update docs later on easily 👍  @dprotaso feel free to veto :)

**Pull request merged**